sdcflows/interfaces/bspline.py

```diff
@@ -0,0 +1,249 @@
+from nipype.utils.filemanip import fname_presuffix
+from nipype.interfaces.base import (
+    BaseInterfaceInputSpec,
+    TraitedSpec,
+    File,
+    traits,
+    SimpleInterface,
+    InputMultiObject,
+    OutputMultiObject,
+)
+
+
+DEFAULT_ZOOMS_MM = (40.0, 40.0, 20.0)       # For human adults (mid-frequency), in mm
+DEFAULT_LF_ZOOMS_MM = (100.0, 100.0, 40.0)  # For human adults (low-frequency), in mm
+DEFAULT_HF_ZOOMS_MM = (16.0, 16.0, 10.0)    # For human adults (high-frequency), in mm
+
+
+class _BSplineApproxInputSpec(BaseInterfaceInputSpec):
+    in_data = File(exists=True, mandatory=True, desc="path to a fieldmap")
+    in_mask = File(exists=True, mandatory=True, desc="path to a brain mask")
+    bs_spacing = InputMultiObject(
+        [DEFAULT_ZOOMS_MM],
+        traits.Tuple(traits.Float, traits.Float, traits.Float),
+        usedefault=True,
+        desc="spacing between B-Spline control points",
+    )
+    ridge_alpha = traits.Float(
+        1e-4, usedefault=True, desc="controls the regularization"
+    )
+    recenter = traits.Enum(
+        "mode", "median", "mean", "no", usedefault=True,
+        desc="strategy to recenter the distribution of the input fieldmap"
+    )
+    extrapolate = traits.Bool(True, usedefault=True,
+                              desc="generate a field, extrapolated outside the brain mask")
+
+
+class _BSplineApproxOutputSpec(TraitedSpec):
+    out_field = File(exists=True)
+    out_coeff = OutputMultiObject(File(exists=True))
+    out_error = File(exists=True)
+    out_extrapolated = File()
+
+
+class BSplineApprox(SimpleInterface):
+    """
+    Approximate the field to smooth it removing spikes and extrapolating beyond the brain mask.
+
+    Examples
+    --------
+
+    """
+
+    input_spec = _BSplineApproxInputSpec
+    output_spec = _BSplineApproxOutputSpec
+
+    def _run_interface(self, runtime):
+        from sklearn import linear_model as lm
+
+        # Load in the fieldmap
+        fmapnii = nb.load(self.inputs.in_data)
+        data = fmapnii.get_fdata(dtype="float32")
+        oriented_nii = canonical_orientation(fmapnii)
+        oriented_nii.to_filename("data.nii.gz")
+        mask = nb.load(self.inputs.in_mask).get_fdata() > 0
+        bs_spacing = [np.array(sp, dtype="float32") for sp in self.inputs.bs_spacing]
+
+        # Recenter the fieldmap
+        if self.inputs.recenter == "mode":
+            from scipy.stats import mode
+            data -= mode(data[mask], axis=None)[0][0]
+        elif self.inputs.recenter == "median":
+            data -= np.median(data[mask])
+        elif self.inputs.recenter == "mean":
+            data -= np.mean(data[mask])
+
+        # Calculate spatial location of voxels, and normalize per B-Spline grid
+        mask_indices = np.argwhere(mask)
+        fmap_points = (
+            oriented_nii.affine.astype("float32") @ (
+                np.vstack((mask_indices.T, np.ones((1, mask_indices.shape[0],), dtype=int)))
+            )
+        )[:3].T
```

Comment:
Is this just [`apply_affine`](https://github.com/nipy/nibabel/blob/591a1bd9cfb1d39b6834b5cf432fcbc8a76e3a7b/nibabel/affines.py#L17-L84)?

```suggestion
        fmap_points = nb.affines.apply_affine(oriented_nii.affine.astype("float32"), mask_indices)
```

Might need to add some `.T`s in there...